iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/gcp/GcpPrivilegePolicyCheckPart.java

```diff
@@ -0,0 +1,123 @@
+
+  private final Map<String, AttributeTree> policyReferences = new HashMap<>();
+
+  @Override
+  public void initialize(InitContext init) {
+    super.initialize(init);
+    init.register(FileTree.class, (ctx, tree) -> new PolicyReferenceCollector().scan(new TreeContext(), tree));
+  }
+
+  @Override
+  protected void provideResource(CheckContext ctx, BlockTree blockTree) {
+    super.provideResource(ctx, blockTree);
+    if (isData(blockTree)) {
+      ResourceSymbol data = ResourceSymbol.fromPresent(ctx, blockTree);
+      if ("google_iam_policy".equals(data.type)) {
+        checkPolicy(data);
+      }
+    }
+  }
+
+  private static boolean isData(BlockTree blockTree) {
+    return "data".equals(blockTree.key().value());
+  }
+
+  private void checkPolicy(ResourceSymbol data) {
+    AttributeTree reference = policyReferences.get(formatReference(data));
+    if (reference != null) {
+      SecondaryLocation secondary = new SecondaryLocation(reference, SECONDARY_MESSAGE);
+      data.blocks("binding").forEach(
+        binding -> binding.attribute("role")
+          .reportIf(matchesPattern(".*(?:admin|manager|owner|superuser).*"), POLICY_MESSAGE, secondary));
+    }
+  }
+
+  private static String formatReference(ResourceSymbol policy) {
+    return String.format("data.google_iam_policy.%s.policy_data", policy.name);
+  }
+
+  @Override
+  protected void registerResourceConsumer() {
+    register(List.of("google_project_iam_binding", "google_organization_iam_binding",
+        "google_service_account_iam_binding", "google_folder_iam_binding"),
+      resource -> checkRole(resource, POLICY_MESSAGE));
+
+    register(List.of("google_project_iam_member", "google_organization_iam_member",
+        "google_service_account_iam_member", "google_folder_iam_member"),
+      resource -> checkRole(resource, MEMBER_MESSAGE));
+  }
+
+  private void checkRole(ResourceSymbol resource, String message) {
+    resource.attribute("role")
+      .reportIf(matchesPattern(".*(?:admin|manager|owner|superuser).*"), message);
+  }
+
+  private class PolicyReferenceCollector extends TreeVisitor<TreeContext> {
+
+    private final Set<String> relevantResources = Set.of("google_project_iam_policy", "google_organization_iam_policy",
+      "google_service_account_iam_policy", "google_folder_iam_policy");
+
+    public PolicyReferenceCollector() {
+      policyReferences.clear();
+      register(BlockTree.class, (ctx, tree) -> {
+        if (isResource(tree) && relevantResources.contains(getResourceType(tree))) {
+          collectReference(tree);
+        }
+      });
+    }
+
+    private void collectReference(BlockTree tree) {
+      PropertyUtils.get(tree, "policy_data", AttributeTree.class)
+        .filter(policyData -> policyData.value().is(TerraformTree.Kind.ATTRIBUTE_ACCESS))
+        .ifPresent(policyData -> policyReferences.put(TerraformUtils.attributeAccessToString((AttributeAccessTree) policyData.value()), policyData));
```

Comment:
I've lost it here, I am afraid.
Isn't there a simpler way to express this ?
Couple of static imports (for `TerraforTree.Kind.ATTRIBUTE_ACCESS` & `TerraformUtils.attributeAccessToString`) would certainly help a bit, but beyond that?